crates/test/cometbft-launcher/src/lib.rs

```diff
@@ -0,0 +1,336 @@
+            rpc_listen_addr,
+        })
+    }
+}
+
+/// Returns the cometbft rpc listenaddress by sending a HTTP GET using the /status endpoint.
+async fn find_rpc_listen_address(process: &Child) -> anyhow::Result<SocketAddr> {
+    let pid = process
+        .id()
+        .context(
+            "cometbft process did not have a PID; this can only happen if it already completed",
+        )?
+        .try_into()
+        .expect("failed converting a u32 PID to i32; this should never fail in practice");
+
+    let rpc_listen_addr = tryhard::retry_fn(move || async move {
+        let addrs = tokio::task::spawn_blocking(move || enumerate_local_tcp_socket_addrs(pid))
+            .await
+            .context("task panicked looking for ports cometbft is listening on")?
+            .context("failed looking for local ports")?;
+        let client = reqwest::Client::new();
+        for addr in addrs {
+            let Ok(response) = client
+                .get(format!("http://{addr}/status"))
+                .timeout(Duration::from_millis(100))
+                .send()
+                .await
+            else {
+                continue;
+            };
+            if response.status().is_success() {
+                return Ok(addr);
+            }
+        }
+        Err(anyhow::anyhow!(
+            "no cometbft tcp socket found that responds to /status"
+        ))
+    })
+    .retries(9)
+    .fixed_backoff(Duration::from_secs(1))
+    .await
+    .context("finding local tcp addresses failed after 10 attempts")?;
+
+    Ok(rpc_listen_addr)
+}
+
+/// Returns all TCP socket addresses of the process identified by `pid`.
+#[cfg(target_os = "macos")]
+fn enumerate_local_tcp_socket_addrs(pid: i32) -> anyhow::Result<Vec<SocketAddr>> {
+    use std::net::{
+        IpAddr,
+        Ipv4Addr,
+        Ipv6Addr,
+    };
+
+    use anyhow::Error;
+    use libc::{
+        AF_INET,
+        AF_INET6,
+    };
+    use libproc::libproc::{
+        bsd_info::BSDInfo,
+        file_info::{
+            pidfdinfo,
+            ListFDs,
+            ProcFDType,
+        },
+        net_info::{
+            SocketFDInfo,
+            SocketInfoKind,
+        },
+        proc_pid::{
+            listpidinfo,
+            pidinfo,
+        },
+    };
+
+    let info = pidinfo::<BSDInfo>(pid, 0)
+        .map_err(Error::msg)
+        .context("failed getting process info")?;
+    let mut local_addresses = vec![];
+    for fd in listpidinfo::<ListFDs>(pid, info.pbi_nfiles as usize)
+        .map_err(Error::msg)
+        .context("failed listing file descriptors associated with pid")?
+    {
+        if let ProcFDType::Socket = fd.proc_fdtype.into() {
+            let socket = pidfdinfo::<SocketFDInfo>(pid, fd.proc_fd)
+                .map_err(Error::msg)
+                .context("failed getting info on socket fd")?;
+            if let SocketInfoKind::Tcp = socket.psi.soi_kind.into() {
+                // unsafe ok because union is discriminated by soi_kind
```

Comment:
what does this mean exactly?

Reply:
These are all C structs read into Rust. C does not have a language native concept of a tagged union, so you need to manually keep track of what you store in your union. In Rust the equivalent to an `enum` is actually something like this:
```rust
struct MyTaggedUnion {
    tag: u8,
    the_union: MyUnion,
}

union MyUnion {
    Number(u64),
    Str(String),
}

const THE_NUMBER: u8: 0;
const THE_STRING: u8: 1;

# an instance of my tagged union
const demo: MyTaggedUnion = MyTaggedUnion {
    tag: THE_NUMBER,
    the_union: MyUnion::Number(42),
};
```

So I am discriminating what's in `MyTaggedUnion` by the `tag` field.